Feature: CMake build

mahrud opened the pull request:
Here's the PR that closes #1080. I'll update this description to include a summary of changes soon.

DanGrayson commented:
I tried building as usual (with make) and encountered this:
```
make[3]: *** No rule to make target '/Users/dan/src/M2/M2.git/M2/Macaulay2/e/../m2/statuscodes', needed by '/Users/dan/src/M2/M2.git/M2/Macaulay2/e/TAGS'.  Stop.
make[3]: Leaving directory '/Users/dan/src/M2/M2.git/M2/BUILD/dan/builds.tmp/einsteinium-1190/Macaulay2/e'
make[2]: *** [Makefile:14: all-in-e] Error 2
```

DanGrayson commented on M2/INSTALL-CMake.md:
````diff
@@ -0,0 +1,327 @@
+
+A quick build involves the following steps:
+```
+git clone https://github.com/mahrud/M2.git -b feature/cmake
````
This should point to the main repository, master branch.

DanGrayson commented on M2/INSTALL-CMake.md:
````diff
@@ -0,0 +1,327 @@
+### Building Macaulay2
+1. Clone Macaulay2 and switch to branch `feature/cmake`:
+```
+git clone https://github.com/mahrud/M2.git -b feature/cmake
````
This, too.

DanGrayson commented on M2/INSTALL-CMake.md:
````diff
@@ -0,0 +1,327 @@
+```
+ninja build-libraries
+```
+Note that this target **must** be built separately, before proceeding to `M2-binary`.
````
Couldn't ninja be told to do this as a prerequisite for building M2-binary?

mahrud replied:
It could, but that would be an anti-pattern. The idea is that ninja should **not** build libraries by default, because the default task is building M2. If a package isn't found, then the user should get an error.

The `build-libraries` target is a "[superbuild](https://blog.kitware.com/cmake-superbuilds-git-submodules/)" target, so it needs to come before the "build" step.

mahrud replied:
This solves a bug with the comfigure/make system, which is that flint and a few other libraries are built every time you run make, and flint takes a long time to build, so you either need to know exactly which directories need to be rebuilt or wait a long time.

DanGrayson replied:
Flint needing to be rebuilt is not a bug, it's a necessity -- there's no distribution out there yet that will work with our code.  What do you do about that?

mahrud replied:
That's not what I meant. We also build Flint with `ninja build-flint`. But with configure/make, every time you run make Flint is recompiled, even if it is already compiled and exists in the same place.

DanGrayson replied:
I think not.  Look at these files:
```
  /home/dan/src/M2/M2/BUILD/dan/builds.tmp/debian-development/libraries/flint:
  total used in directory 24 available 16117552
  drwxrwxr-x  3 dan dan  4096 May 23 09:22 .
  drwxrwxr-x 37 dan dan  4096 May 23 09:22 ..
  drwxrwxr-x 82 dan dan 12288 May 22 19:05 build
  -rw-rw-r--  1 dan dan     0 May 22 22:20 .checked-20200414
  -rw-rw-r--  1 dan dan     0 May 22 19:05 .compiled-20200414
  -rw-rw-r--  1 dan dan     0 May 22 18:51 .configured-20200414
  -rw-rw-r--  1 dan dan     0 May 22 19:05 .installed-20200414
  -r--r--r--  1 dan dan  1403 May 23 09:22 Makefile
  -rw-rw-r--  1 dan dan     0 May 22 18:51 .patched-20200414
```
We check for the presence of those various files of 0 length to decide whether something has been done already, so we don't do it again.

(That wrinkle hasn't been implemented yet for the submodules, which can be annoying.)

mahrud replied:
Sure, but the `libraries` directory is very heavy, and I often run `make clean -C libraries` after the libraries and programs are installed in `usr-host`. Or I might want to make symlink from `usr-host` to the `usr-host` in another build directory in order to save time and space. In this case, configure doesn't detect that we've already built the right version of Flint. Forcing ninja to run `build-libraries` would cause it to rebuild the libraries, as the current build system does, even though they already exist in `usr-host`.

mahrud replied:
Sorry I wasn't clear about "exists the same place" earlier. Shall I mark this conversation as resolved?

DanGrayson replied:
Ah, I've never considered cleaning in `libraries` after building them.

And yes, flint offers no version number we can test for.

Resolved.

DanGrayson replied:
One simple way to avoid rebuilding libraries and submodules is to run make in `Macaulay2/.`.  Or run `make all-in-Macaulay2` at top level.

mahrud replied:
`flint/flint.h` includes the following:
```
#define FLINT_VERSION "2.6.0"
```
The standard practice with CMake is to use regex matching with header files to detect version, rather than compiling a program to extract version. Here is an example:
https://github.com/mahrud/M2/blob/feature/cmake/M2/cmake/FindFlint.cmake#L35-L46

DanGrayson commented on M2/Macaulay2/bin/main.cpp:
```diff
@@ -0,0 +1,329 @@
+ * these two macros affect the definition of GC_INIT, but have to
+ * appear before the include directives, in order to take effect
+ */
+#define GC_FREE_SPACE_DIVISOR 12
```
These lines are in M2/include/M2/gc-include.h, so should not be here.

DanGrayson commented on M2/Macaulay2/bin/main.cpp:
```diff
@@ -0,0 +1,329 @@
+#endif
+
+  exit(0);
+  return NULL;
```
This line should not be needed, provided "exit" is given the "noreturn" attribute.  But to avoid
all doubt, replace the pair of lines by "return 0".  Returning NULL is inappropriate, because NULL is for pointers, and main is to return an int.

mahrud replied:
This is from `interpFunc` here:

mahrud replied:
This is simply moved from `interpFunc` here:
https://github.com/Macaulay2/M2/blob/8a473f501318a6d505958ad6332e6618e5f16948/M2/Macaulay2/d/M2lib.c#L609-L612
Is there a reason why `interpFunc` returns a `void*` pointer?

DanGrayson replied:
Good point.  And I was wrong about it returning an int -- I must have thought the lines were in `main()`.  Gary Furnish must have made it return a pointer so it could be handed to `createThreadTask`.

Leave it as it is for now.  A future experiment would involve removing the "return" and leaving the call to "exit".

DanGrayson commented on M2/Macaulay2/d/CMakeLists.txt:
```diff
@@ -0,0 +1,199 @@
+
+# CAUTION: order is important in this list
+# TODO: can there be #ifdefs in D sources?
+set(DLIST
```
Could we arrange for lists like this to be shared by both build systems?  That would prevent certain sorts of future errors.

mahrud replied:
The way the `d` directory is currently built, I don't know of a clean way to do this, because building some `.d` sources depends on flags (e.g. XML, Python, or MySQL). However, if `scc1` supported `#ifdef`s in its syntax, then we could put the list in a file and build everything in it, knowing that the compiler would skip the content if a flag is not set.

DanGrayson replied:
For that reason, you've convinced me that it would be good to preprocess the D files.  The way to do that is this: make a shell script called `c/scc`.  It would run `cpp` on the file first and then run `c/scc1`.  The code of `scc1` would be modified to pay attention to the lines left behind `cpp` : a line such as `# 11 "foo.d" 2` would tell it that the next line (check that) is line 11 from the file "foo.d".  We need that information for giving proper line numbers in error messages, and we don't want to be off even by 1.

The reason the "1" is in the name "scc1" is to allow for this, I've just remembered.

Feel free to make this comment thread into an issue.

mahrud replied:
Made it an issue in #1197. Setting this thread as resolved for now.

DanGrayson replied:
@mikestillman  may like to do this for the `e` directory.  If there are no conditionals, how would that be done?  (Not saying to do it now ... )

mahrud replied:
This would be a little complicated right now, since there are multiple classes of sources. Perhaps after some engine refactoring it'll be nicer.

DanGrayson commented on Makefile:
```diff
@@ -0,0 +1,31 @@
+# this is a file to create build directories here
```
We already have a place for build directories: `M2/BUILD`.  And instead of letting the user make them, just make them and populate each one with appropriate files for the user to use.  The names could be `M2/BUILDS/cmake/debug`, etc.

mahrud replied:
@mikestillman do you need this file?

mikestillman replied:
@mahrud Please remove that top level Makefile.  I no longer use it.

DanGrayson commented on bugs/mahrud/atomic_test/test.cpp:
```diff
@@ -0,0 +1,39 @@
+#include <vector>
```
These files don't report bugs, so they don't belong in the `bugs` directory, which we want to get rid of.  They also don't seem to have anything to do with Macaulay2, so why are they here?

mahrud replied:
This is a test, but some files in `bugs/mahrud` are useful scripts for debugging. There didn't seem to be a better place to share them. Should I move them to `M2/BUILD/mahrud` instead?

DanGrayson replied:
That would be good, so we  can work on eliminating `bugs` entirely.  As you have remarked, those old bug reports are mostly useless.  There's just one of them that I want to save, because it consists of test procedures that should be moved to something like `M2/INSTALL-manual-testing`.

mahrud replied:
Moved. I've been using the `atomic`, `thread_local`, and `pthread` tests to check those capabilities on different machines, and I think are useful for making sure a bug on a specific machine isn't caused by some architecture dependent behavior.

DanGrayson replied:
Yes, it's worthwhile to save those somewhere ...

DanGrayson commented on bugs/mahrud/trace_test.m2:
```diff
@@ -0,0 +1,22 @@
+debug Core
```
This file is related to Macaulay2, but it doesn't report a bug.  Is it a test?  An experiment?

mahrud replied:
It's a test of the PROFILING option, but not one that could be moved to `Macaulay2/tests` since it requires a build option. Where should I move it?

DanGrayson replied:
Let's devise a way for it to tell whether the test can be run, so the file can be in `Macaulay2/tests`.  Maybe we could add some things to `version`.

By the way, could it ever fail?  Does it check or verify that any results are correct?  If not, in what sense is it a test?

mahrud replied:
They can fail if the stack trace shows too many things or not enough things, or if it crashes. The profiling feature is still rather experimental. When it is more ironed out I will add better tests, but that would be a future PR.

DanGrayson commented:
I think you need this change, too:
```
diff --git a/M2/Macaulay2/e/Makefile.in b/M2/Macaulay2/e/Makefile.in
index f28fe0a3d..16503d801 100644
--- a/M2/Macaulay2/e/Makefile.in
+++ b/M2/Macaulay2/e/Makefile.in
@@ -26,8 +26,7 @@ TAGS tags: @srcdir@/TAGS
 @srcdir@/TAGS: Makefile.in @srcdir@/*.hpp @srcdir@/*.h @srcdir@/*.cpp @srcdir@/*.c \
 		 @srcdir@/f4/*.hpp @srcdir@/f4/*.cpp \
 		 @srcdir@/bibasis/*.hpp @srcdir@/bibasis/*.cpp \
-		 @srcdir@/unit-tests/*.cpp \
-		 @srcdir@/../m2/statuscodes
+		 @srcdir@/unit-tests/*.cpp
 	@ echo " ** making TAGS file"
 	@ cd @srcdir@ && @ETAGS@ -o TAGS $(patsubst @srcdir@/%, %, $^)
```

DanGrayson commented:
And this:
```
diff --git a/M2/Macaulay2/m2/Makefile.in b/M2/Macaulay2/m2/Makefile.in
index e5076d4de..5bc185af7 100644
--- a/M2/Macaulay2/m2/Makefile.in
+++ b/M2/Macaulay2/m2/Makefile.in
@@ -136,7 +136,7 @@ clean::; rm -f profile
 @pre_packagesdir@ :; @INSTALL@ -d $@
 @pre_packagesdir@/Core  : ; @INSTALL@ -d $@
 @pre_packagesdir@/Core/% : % ; @INSTALL_DATA@ $^ $@
-FILES_TO_COPY := $(DUMPEDM2FILES) loadsequence installedpackages statuscodes
+FILES_TO_COPY := $(DUMPEDM2FILES) loadsequence installedpackages
 all:@pre_packagesdir@ @pre_packagesdir@/Core $(addprefix @pre_packagesdir@/Core/, $(FILES_TO_COPY))
 clean::; rm -rf cache *.obj *.exe runexample temp.data distfiles loads.m2 mpcodes.m2 core gmon.out
 distclean: clean; rm -f Makefile version.m2
```

mahrud commented:
By the way, also take a look at these accompanying PRs:
- https://github.com/Macaulay2/memtailor/pull/5
- https://github.com/Macaulay2/mathic/pull/5
- https://github.com/Macaulay2/mathicgb/pull/16
- https://github.com/Macaulay2/flint2/pull/1

I've also updated and renamed the `gtest` submodule `googletest`, which is the correct name of its repository, and updated `bdwgc` to the upstream version, which supports building with CMake.

mahrud commented:
@DanGrayson anything else needed here?

DanGrayson commented:
Maybe not.  Could you resolve the conflicts?  Then I'll take a look.

mahrud commented:
Didn't realize the docker branch was going to be merged, but finally got around to making the docker changes I've been meaning to make.

DanGrayson commented:
Mahrud, is this pull request now ready to be considered again?

mahrud commented:
Yes, I resolved the conflicts, as you asked.

DanGrayson commented on .gitattributes:
```diff
@@ -0,0 +1,3 @@
+*.m2 linguist-language=Macaulay2
```
Do we have to do something else now so that files such as those at https://github.com/Macaulay2/M2/blob/master/M2/Macaulay2/m2/debugging.m2 get highlighted beautifully?

mahrud replied:
The PR to add Macaulay2 highlighting to Linguist has been merged, but Linguist needs to make a release before it goes into effect. That should be soon, but ideally we should update the [language grammar](https://github.com/Macaulay2/language-macaulay2/blob/master/grammars/macaulay2.cson) to include recent changes before then.

DanGrayson commented:
Building with autotools and make I get this:
```
+ grep '^.define HAVE_FLINT 1' _config.h
make[3]: *** [../Makefile.library:181: .configured-4.1.1] Error 1
make[3]: Leaving directory '/Users/dan/src/M2/M2.git/M2/BUILD/dan/builds.tmp/einsteinium-1190/libraries/factory'
make[2]: *** [Makefile:7: all-in-factory] Error 2
make[2]: Leaving directory '/Users/dan/src/M2/M2.git/M2/BUILD/dan/builds.tmp/einsteinium-1190/libraries'
make[1]: *** [GNUmakefile:210: all-in-libraries] Error 2
make[1]: Leaving directory '/Users/dan/src/M2/M2.git/M2/BUILD/dan/builds.tmp/einsteinium-1190'
make: *** [Makefile.include:17: all] Error 2
```

Factory's config.log shows this:
```
conftest.c:31:26: error: use of undeclared identifier 'flint_version'
        int main () { if ((int) flint_version[0] < 2) return -1; else return 0; }
                                ^
1 error generated.
```

That undefined variable seems to have arisen from a change in this pull request:
```
--- a/M2/libraries/factory/patch-4.1.1
+++ b/M2/libraries/factory/patch-4.1.1
@@ -21,3 +21,15 @@ diff -ur /Users/dan/src/M2/M2-Macaulay2/M2/BUILD/dan/builds.tmp/einsteinium-mast
    //Off(SW_USE_NTL_SORT);
  #endif
    On(SW_USE_EZGCD);
+diff -ur ~/Projects/M2/M2/M2/BUILD/mahrud/build/libraries/factory/build/m4/flint-check.m4 m4/flint-check.m4
+--- /home/mahrud/Projects/M2/M2/M2/BUILD/mahrud/build/libraries/factory/build/m4/flint-check.m4	2018-02-15 08:02:18.000000000 -0600
++++ m4/flint-check.m4	2020-05-26 15:54:26.373657808 -0500
+@@ -69,7 +69,7 @@
+ 	[
+ 	AC_TRY_RUN(
+ 	[#include <flint/flint.h>
+-	int main () { if ((int) version[0] < 2) return -1; else return 0; }
++	int main () { if ((int) flint_version[0] < 2) return -1; else return 0; }
+ 	],[
+ 	flint_found="yes"
+ 	break
```

DanGrayson commented:
Currently `libraries/flint/Makefile.in` is set to an older version of flint, April 20, whereas `submodules/flint2`  points to a modern version of flint, May 26, patched by you.

mahrud commented:
We should update the Makefile version, then. Why not switch to a submodule for that as well?

DanGrayson commented:
Yes, good idea.  And it would be good if both build systems would use the same version of flint.  Then I won't have to futz with uploading tarfiles, too.

PS: We should have a way to tell whether a running version of M2 was built with autoconf+make or with cmake+ninja or with cmake+make, etc., by looking in `version`.  What would you suggest?  At least, for cmake builds, `version#"configure arguments"` should be absent or a string of length 0.  Is it?  And for cmake builds, it would be good to be able to observe optional arguments to cmake.  Are there any?

mahrud commented:
Can you make that change? I don't know how autotools handles submodule libraries.

There are plenty of optional arguments for cmake (see [INSTALL-CMake.md](https://github.com/mahrud/M2/blob/feature/cmake/M2/INSTALL-CMake.md)), but I don't think that's a good idea. I have actually removed "configure arguments", since any change in `config.h` has a cascade effect on basically all sources, and everything recompiling for something like `--enable-rpm` being removed is a waste of time and resources. But if you'd like we can replace it with "build tool" which can be either "autotools" or "cmake" and doesn't change as often.

What is the benefit of knowing how the build environment was configured?

mahrud commented:
We can have that discussion another time. I added "configure arguments" again, and for now on CMake build it just has the build type, generator, and whether CPU vectorization is used.

DanGrayson commented:
> Can you make that change? I don't know how autotools handles submodule libraries.

Sure, but not soon.  Make an issue, assign me, and meanwhile undo that change to the factory patch file.

> 
> There are plenty of optional arguments for cmake (see [INSTALL-CMake.md](https://github.com/mahrud/M2/blob/feature/cmake/M2/INSTALL-CMake.md)), but I don't think that's a good idea. I have actually removed "configure arguments", since any change in `config.h` has a cascade effect on basically all sources, and everything recompiling for something like `--enable-rpm` being removed is a waste of time and resources. But if you'd like we can replace it with "build tool" which can be either "autotools" or "cmake" and doesn't change as often.

Okay, sure.   

> 
> What is the benefit of knowing how the build environment was configured?

If there's a bug reported by a user, reproducing the bug at our end might depend on building it the same way.  So any build options that affect the code should be reported in `version`.  Not --enable-rpm, of course, which I should also get rid of, as you suggest.

mahrud commented:
> Sure, but not soon. Make an issue, assign me, and meanwhile undo that change to the factory patch file.

That would break factory working with the new flint .. I'll think of something else.

> If there's a bug reported by a user, reproducing the bug at our end might depend on building it the same way. So any build options that affect the code should be reported in `version`. Not --enable-rpm, of course, which I should also get rid of, as you suggest.

Wouldn't the git commit suffice for that? And if they compiled it a different way, we could ask the user to also tell us what config arguments they used.

DanGrayson commented:
The git commit (if it's not dirty), the way of building it, the build options that affect the code, plus the version numbers of the libraries in use, and the list of libraries that are built by us, might suffice.  And there's no sense asking the user -- they might have gotten it from us or from someone else or forgotten or not know how to tell.

mahrud commented:
Turns out the flint version check in factory has been wrong for years. It is comparing the character "2" with the integer 2 ... I fixed it a different way that should be compatible now.

I also updated the submodules, so before building these should probably be merged:

- https://github.com/Macaulay2/memtailor/pull/6
- https://github.com/Macaulay2/mathic/pull/6
- https://github.com/Macaulay2/mathicgb/pull/17
- https://github.com/Macaulay2/flint2/pull/2

After this, I think everything should work.

mahrud commented:
This should close #1206 for CMake, but I don't know where's the right place to add build flags for submodules using autotools.

DanGrayson commented:
> This should close #1206 for CMake, but I don't know where's the right place to add build flags for submodules using autotools.

Yes, leave it open.

DanGrayson commented:
> > Sure, but not soon. Make an issue, assign me, and meanwhile undo that change to the factory patch file.
> 
> That would break factory working with the new flint .. I'll think of something else.
> 

It won't break it today, but later on, so don't worry.

mahrud commented:
> It won't break it today, but later on, so don't worry.

Sorry, I meant it breaks building factory from CMake, since it builds the more recent flint. Specifically, linking with static flint and factory wasn't working before I changed that patch.

Anyway, that should be fixed now.

mahrud commented:
@DanGrayson this is ready for you to try out again. There was a mismatch in fflas-ffpack commit earlier that broke the autotools build, but I just changed it to match master and it works again. Also my Singular PR was merged this morning so I updated the commit and change the module to follow the upstream.

mahrud commented:
This is now tested on Catalina. gtests for memtailor, mathic, mathicgb have a strange issue, but beside that everything worked after setting this cflag:
```
--sysroot=`xcrun --show-sdk-path`
```

DanGrayson commented:
This new feature doesn't quite work:

```
+ ./M2 --no-readline --print-width 92
Macaulay2, version 1.15.1.0
--loading configuration for package "FourTiTwo" from file /Users/dan/Library/Application Support/Macaulay2/init-FourTiTwo.m2
--loading configuration for package "Topcom" from file /Users/dan/Library/Application Support/Macaulay2/init-Topcom.m2
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : debug Core

i2 : spin 10000
  C-c C-c  C-c C-c
Exist (y=yes/n=no/a=abort/b=backtrace)? a
returning to top level
../../../../Macaulay2/m2/startup.m2.in:21:13:(0): error: assignment to protected global variable 'assert', originally defined at ../../../../Macaulay2/m2/startup.m2.in:21:6:(0):
../../../../Macaulay2/m2/startup.m2.in:21:13:(0): --entering debugger (type help to see debugger commands)
../../../../Macaulay2/m2/startup.m2.in:21:13-23:25: --source code:
     assert = x -> (
          if class x =!= Boolean then error "assert: expected true or false";
          if not x then error "assertion failed");
```

We've discussed this before.  This would fix it:

```
diff --git a/M2/Macaulay2/m2/startup.m2.in b/M2/Macaulay2/m2/startup.m2.in
index 3fd596cd1..7563743f1 100644
--- a/M2/Macaulay2/m2/startup.m2.in
+++ b/M2/Macaulay2/m2/startup.m2.in
@@ -14,8 +14,6 @@ if gotarg "--int"  then handleInterrupts = false
 
 firstTime := class PackageDictionary === Symbol
 
-firstTime = true					    -- debugging
-
 if firstTime then (
      -- we do this bit *before* "debug Core", so that Core (the symbol, not the package), which may not be there yet, ends up in the right dictionary
      assert = x -> (
@@ -616,6 +614,7 @@ homeDirectory = getenv "HOME" | "/"
 
 initcurrentlayout()
 
+if firstTime then
 copyright = copyright | newline | "For licenses see " | prefixDirectory | currentLayout#"program licenses"
 
 path = (x -> select(x, i -> i =!= null)) deepSplice {
```

mikestillman commented:
I get the following warning during `ninja M2-core`
```
[20/331] Generating d/TAGS file
distributed-packages.h: No such file or directory
```

mahrud commented on M2/Macaulay2/docs/conf.py:
```diff
@@ -0,0 +1,54 @@
+# -- Project information -----------------------------------------------------
+
+project = 'Macaulay2 Internal Documentation'
+copyright = '2020, The Macaulay2 Authors'
```
How is it undefined?
Regardless, that config is for Sphinx, which we're not yet settled on and doesn't contain any information. You can ignore it for now. It'll either be removed or improved.

DanGrayson replied:
It's undefined in that there is no definition offered.

mahrud commented on M2/Macaulay2/d/Makefile.in:
```diff
@@ -210,27 +206,30 @@ ifeq (@HAVE_WFRAME_ADDRESS@,yes)
 endif
 
+# flags needed to sidestep system regex
+scclib.o : CPPFLAGS += -Dre_compile_fastmap=M2_re_compile_fastmap -Dre_compile_pattern=M2_re_compile_pattern -Dre_match=M2_re_match -Dre_match_2=M2_re_match_2 -Dre_search=M2_re_search -Dre_search_2=M2_re_search_2 -Dre_set_registers=M2_re_set_registers -Dre_set_syntax=M2_re_set_syntax -Dregcomp=M2_regcomp -Dregerror=M2_regerror -Dregexec=M2_regexec -Dregfree=M2_regfree
```
Yes; I have a branch where regex.h from system is used instead of Macaulay2/regex. Putting the macros in Makefile allows having an option to do that.

DanGrayson replied:
It would be better to have one command line macro definition, such as "M2_REGEX", to trigger the others, to keep the command line less crowded, and to have that long list of names in just one place, rather than in two places (a make file and a cmake file).

mahrud replied:
Can we postpone this to #1092? The faster this is merged, the faster I can get to switching to a better regex engine.

DanGrayson replied:
Okay, but don't forget.

mahrud commented on M2/Macaulay2/m2/html.m2:
```diff
@@ -464,24 +464,47 @@ describeReturnCode = r -> (
+     usermode = if usermode === false or usermode === null then defmode else usermode; -* the default *-
+     usermode = if usermode < 0 then (1<<64) + defmode - usermode else usermode; -* useful for changing the default *-
+     -* returns (" "|arg) if the (m)th bit is set *-
+     readmode := (m, arg) -> if usermode & (1 << m) != 0 then " " | arg else "";
```
It does not change the top level behavior, as it is backwards compatible; i.e. true, false, and null values are processed exactly as before.

mahrud replied:
In a future PR I'll move this function out of html.m2, and at that point I might document and export it. For now, it serves a very useful purpose for running tests with arbitrary command line options (e.g. disabling threading, readline, etc.) without having an option table with 30 different values.

DanGrayson replied:
> It does not change the top level behavior, as it is backwards compatible; i.e. true, false, and null values are processed exactly as before.

It does change the top level behavior, even though it is backward compatible, as you say.

DanGrayson replied:
> In a future PR I'll move this function out of html.m2, and at that point I might document and export it. For now, it serves a very useful purpose for running tests with arbitrary command line options (e.g. disabling threading, readline, etc.) without having an option table with 30 different values.

We don't have any tests needing that, do we?

mahrud replied:
I do. For ctests, I don't want the output to be captured in a file. This way `ctest -V` can show you the output and highlight where an error occurred based on the given regex.

DanGrayson replied:
What's wrong with the standard way?: Signal an error and save the output file so you can look at it later.

mahrud replied:
I can think of a few reasons:
- why keep an extra file? I can't find a unit testing framework that logs all outputs, all the time.
- if I'm looking for a regex in the output to interrupt computation, I have to see the output directly (and no, convoluted bash pipes are not the answer)
- what makes `--no-randomize --silent --print-width 77 --stop --int --no-readline` "standard"?
- what if I need to test readline, test without threads/preload/setup/etc, or test the new `--webapp`?

mahrud replied:
All that is beside the point: I wanted to take advantage of ctest features to [debug this](https://github.com/Macaulay2/M2/pull/1096#issuecomment-619660197), so made a backwards compatible change which is intentionally not documented because an unexported `runFile` function in `html.m2` with 9 inputs is definitely not the right place to run tests, so I see it as a temporary patch until there's a proper unit testing framework in, say, `testing.m2`.

DanGrayson replied:
We need the two build systems to run the same tests the same way, and the change to the UserMode option is not acceptable.  If you need a new way to run tests, we can jointly come up with a better way.  For example, what if we had some test files that are simply shell scripts?  Such a shell script could run M2 with any command line options.

mahrud replied:
> We need the two build systems to run the same tests the same way

Why? The new build system is not designed to simply mimic the old build system. It is designed to make development easier, and part of that is making it easier to run tests. If I want to run only `tests/engine/raw-hilb.m2` 100 times in parallel, I can just do `ctest -R engine/raw-hilb --repeat-until-fail 100 -j4`.

> the change to the UserMode option is not acceptable

What does the change to the UserMode option change for the configure build? Nothing at all. Why are you against a backwards compatible change which enhances flexibility of running tests?

I could copy the `runFile` function and make `runFile2` which does what I need and write a `check2` method to go with it, but that would just be duplicated code. Would that be better?

> If you need a new way to run tests, we can jointly come up with a better way.

I do. We can make a new issue and deal with that in a different PR.

DanGrayson replied:
> > We need the two build systems to run the same tests the same way
> 
> Why? The new build system is not designed to simply mimic the old build system. It is designed to make development easier, and part of that is making it easier to run tests. If I want to run only `tests/engine/raw-hilb.m2` 100 times in parallel, I can just do `ctest -R engine/raw-hilb --repeat-until-fail 100 -j4`.

What I mean is, that, ignoring bells and whistles, I should be able to issue one command to run all the tests, in either build system, and it should be the same tests, run in the same way.

> 
> > the change to the UserMode option is not acceptable
> 
> What does the change to the UserMode option change for the configure build? Nothing at all. Why are you against a backwards compatible change which enhances flexibility of running tests?

It's an egregious kludge, in that it assigns arbitrary meanings to 18 numbers.  And there's certainly a better way, such as a I used here: https://github.com/Macaulay2/M2/blob/master/M2/Macaulay2/tests/normal/interrupt-handling.m2 , or what I suggested above.  And I can't even find an instance where you're using this new facility.


> 
> I could copy the `runFile` function and make `runFile2` which does what I need and write a `check2` method to go with it, but that would just be duplicated code. Would that be better?

No, that would be even worse, with two new functions with non-intuitive names.

> 
> > If you need a new way to run tests, we can jointly come up with a better way.
> 
> I do. We can make a new issue and deal with that in a different PR.

Good.  Meanwhile, put back the old behavior of UserMode.

mahrud replied:
> I should be able to issue one command to run all the tests, in either build system, and it should be the same tests, run in the same way.

You can run `ninja check-packages` and the exact same thing will be executed; no more and no less. For those in `tests` directory, `ctest -R normal` or `ctest -R engine` run the same tests, no more and no less.

> It's an egregious kludge, in that it assigns arbitrary meanings to 18 numbers.

That's what a bitmask is, and used in a ton of places to make things more readable. Wouldn't you say the entire `html.m2` an "egregious kludge"?

> And I can't even find an instance where you're using this new facility.

It is set [here](https://github.com/mahrud/M2/blob/feature/cmake/M2/Macaulay2/packages/CMakeLists.txt#L40-L41), then incorporated in [`M2_TEST_TEMPLATE`](https://github.com/mahrud/M2/blob/98f5be6f041bfaf488ecd0d68edee192972069f3/M2/Macaulay2/packages/CMakeLists.txt#L70), which is then used [here](https://github.com/mahrud/M2/blob/98f5be6f041bfaf488ecd0d68edee192972069f3/M2/Macaulay2/packages/CMakeLists.txt#L214-L216). It is only used if you want to run individual package tests with `ctest -R check-LocalRings-3` for instance. You can also change it by setting the environment variable `UserMode`, which has been invaluable. I'm happy to explain how it works.

> No, that would be even worse, with two new functions with non-intuitive names.

Glad we agree on something.

> Good. Meanwhile, put back the old behavior of UserMode.

Like I said, the behavior hasn't changed. "UserMode => true", "UserMode => false", and "UserMode => null" do the same exact thing, and nowhere in the code you assign an integer to `UserMode` I don't want to refactor the `m2` directory in this PR.

mahrud replied:
> And there's certainly a better way, such as a I used [here](https://github.com/Macaulay2/M2/blob/master/M2/Macaulay2/tests/normal/interrupt-handling.m2) or what I suggested above.

I'm open to different suggestions that would allow me to modify the arguments on any test, without changing the test or the code. With my suggested method, you can pick and choose the arguments and run any test on any package you want. The suggestions you linked are specific to one test or file.

By the way, I think starting a fresh instance of M2 for every test or example when we have threads is a great waste, but that's for another PR, which we won't get to until we get past this.

DanGrayson replied:
> > It's an egregious kludge, in that it assigns arbitrary meanings to 18 numbers.
> 
> That's what a bitmask is, and used in a ton of places to make things more readable. 

But in this case, you've made things *less* readable.  

Yes, bitmasks can make things more readable.  Here's how.  Suppose you have a function `f` with 5 Boolean arguments.  Then you might call it with `f(0,0,1,0,1)`.  Less cryptic would be `f(false, false, true, false, true)`.  That's still pretty cryptic because no one can remember the order of the arguments.  In a language with optional named arguments, one might be able to type `f(FOO=>true,BAR=>true)`.  That's an improvement, but it's still verbose.  Introducing a bitmask, one can reduce 5 arguments to 1 and write something like `f(0b10100)` or `f((1<<4)|(1<<2))`.  Not very readable, but one can introduce symbols for the appropriate powers of 2 and write `f(FOO|BAR)` instead.  Now we have something readable.  Lots of functions written in low level languages such as C use this method.

In the current pull request, you are taking a step backward in readability, replacing something already symbolic, such as `--notify`, by a hard-to-remember number, namely `11`.  And then in your cmake files you are forced to type this ugly thing:
```
    set(UserMode "(1<<31)+(1<<32)-(1<<64)" # -(1<<8)" # for --no-threads
      CACHE INTERNAL "bitmask for ctest check arguments")
```

mahrud replied:
> This line in html.m2 is particularly unreadable

I totally agree; however, as you always say, the code explains itself: if you read further, each `readmode(n, "--arg")` checks if the n-th bit is set, and if so adds "--arg", and a list of all arguments and the bit associated to them--which has the same order as in startup.m2--comes immediately after. So for instance `(1<<5)` corresponds to `--no-randomize`.

> one can introduce symbols for the appropriate powers of 2 and write f(FOO|BAR) instead. Now we have something readable. Lots of functions written in low level languages such as C use this method.

I'm happy to make this change. The only reason I didn't do that initially is that, when calling `check` from toplevel, we don't have access to those symbols, but at the very least it'll make the line above more readable.

DanGrayson replied:
> > Good. Meanwhile, put back the old behavior of UserMode.
> 
> Like I said, the behavior hasn't changed. "UserMode => true", "UserMode => false", and "UserMode => null" do the same exact thing, and nowhere in the code you assign an integer to `UserMode` I don't want to refactor the `m2` directory in this PR.

Here's an example that shows the behavior has changed.  Here is the behavior before:
```
+ M2 --no-readline -q --print-width 189
Macaulay2, version 1.15
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : check ("FirstPackage", UserMode => 1)
--running test 0 of package FirstPackage on line 54 in file /Applications/Macaulay2-1.15/share/Macaulay2/FirstPackage.m2
--    rerun with: check_0 "FirstPackage"
--making test results in file /var/folders/46/9b86vqxj4hjcngvy7kd7sb140000gn/T/M2-59675-0/0.out
stdio:1:1:(3): error: expected true or false

i2 : 
```
And here is your proposed behavior:
```
+ /Users/dan/src/M2/M2.git/M2/BUILD/dan/builds.tmp/einsteinium-1190/M2 --no-readline --print-width 189 -q
Macaulay2, version 1.15.1.0
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : check ("FirstPackage", UserMode => 1)
--running test 0 of package FirstPackage on line 54 in file ../../../../Macaulay2/packages/FirstPackage.m2
--    rerun with: check_0 "FirstPackage"
--making test results
Macaulay2, version 1.15.1.0
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : exit

i2 : exit

Process M2<1> finished
```
Notice that the running M2 process has offered the user a prompt -- that's why it took two uses of `exit` to get out.

That behavior is going to be hard to document!

DanGrayson replied:
> > one can introduce symbols for the appropriate powers of 2 and write f(FOO|BAR) instead. Now we have something readable. Lots of functions written in low level languages such as C use this method.
> 
> I'm happy to make this change. The only reason I didn't do that initially is that, when calling `check` from toplevel, we don't have access to those symbols, but at the very least it'll make the line above more readable.

No, please don't make that change: it would make things worse.  There's no reason to introduce a synonym for something like `--no-readline`, which is already perfectly readable.  Adding redundant symbols adds an unnecessary burden, and they would only be available in that one cmake file.

DanGrayson replied:
> I'm open to different suggestions that would allow me to modify the arguments on any test, without changing the test or the code. With my suggested method, you can pick and choose the arguments and run any test on any package you want. 

First tell me why this facility would be useful for our users.  I don't see, for example, why it would be useful to enable readline when stdin comes from a file.  It would make no difference.

mahrud commented on M2/Macaulay2/system/supervisor.cpp:
```diff
@@ -1,53 +1,70 @@
+// The maximum number of concurrent threads
+const static unsigned int numCores = std::thread::hardware_concurrency();
+// We allocate between 4 and 16 threads initially, to save trouble with memory allocation.
+const static int maxNumThreads = (numCores < 4) ? 4 : (16 < numCores ? 16 : numCores);
```
Is there a point in having more threads than cores available? They would just block each other. This code prevents allocating more threads than needed on startup (the default 16 threads is just too much).

DanGrayson replied:
You make a good point about blocking.

But is there a point in limiting it to 16 threads even if there are 32 cores?

And the user may know more about the hardware than the routine you are calling.  The documentation even states that the number is 0 if it isn't known: https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency

mahrud replied:
> But is there a point in limiting it to 16 threads even if there are 32 cores?

If the application is actually taking advantage of 32 cores, then no. But we don't have any instances where more than a handful of cores would be used at a time. So for now, allocating for 32 threads, even if you have 32 cores, just eats up memory.

> And the user may know more about the hardware than the routine you are calling. The documentation even states that the number is 0 if it isn't known.

My understanding is that that happens in edge cases, like when the CPU is not fully recognized by the kernel, and certainly not on target CPUs we're building binaries for. If that happens, the default is 4, so much older systems can still have a few threads, and advanced users (e.g. super computer) can patch that code.

mahrud commented on M2/Macaulay2/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,218 @@
+###############################################################################
+## Default resource limits
+# TODO: add memory, virtual memory, and stack size
+# the tests aren't run through bash, so ulimit wouldn't work
```
I don't think ulimits are the best practice for this purpose. For testing in particular, it's better to measure the memory usage and report the amount in addition to a red flag if it goes beyond a certain amount rather than interrupting the process right away.

DanGrayson replied:
Why is it better?

How would you measure it?

mahrud replied:
> Why is it better?

Because the ulimits don't warn you if after a change `help sum` suddenly takes 399MB of ram, even though that's definitely a red flag.

> How would you measure it?

I haven't put it all together yet, and it could be a project for someone else, but perhaps using Valgrind's [massif](https://valgrind.org/docs/manual/ms-manual.html#ms-manual.theoutputgraph)

mikestillman commented:
On ubuntu 18.04, gcc-8, the cmake build (with BUILD_TESTINIG=off, and running perhaps 2-3 commits behind where the branch is now) gives a number of errors in the example for `SumsOfSquares`, `SemidefiniteProgramming`. 

It appears that it cannot execute its command,which I think is `csdp`, although when I look at `csdp`, via `ldd`, it seems fine, so I'm not sure yet what the problem is.

mahrud commented:
@DanGrayson 
- fixed startup.m2.in as you suggested, but see the github issue for the main problem.
@mikestillman 
- fixed the d/TAGS warning
- added tip about M2_HOST_PREFIX to INSTALL-CMake.md
- temporarily disabled tests only for memtailor, mathic, mathicgb, until the mac issues can be resolved. Tests can still be run manually, and they do pass.

mahrud commented:
> On ubuntu 18.04, gcc-8, the cmake build (with BUILD_TESTINIG=off, and running perhaps 2-3 commits behind where the branch is now) gives a number of errors in the example for `SumsOfSquares`, `SemidefiniteProgramming`.
> 
> It appears that it cannot execute its command,which I think is `csdp`, although when I look at `csdp`, via `ldd`, it seems fine, so I'm not sure yet what the problem is.

It's probably because csdp exists on your system. Try `-DBUILD_PROGRAMS=all` then `ninja build-programs`. Once we have a central way for finding programs on the system, this won't be required.

mikestillman commented:
I don't think csdp is on the system (`which csdp` returns with no output) , but it is in the libexec/.../bin where we expect it to  be.

mahrud commented:
@mikestillman They all pass for me on ubuntu. Could you email me or show one failed output of a test? This command would do:
```
ctest -R check-SumsOfSquares -V
```

mikestillman commented:
@mahrud: it is actually failing while making examples (I haven't tried today the tests for these). 

Here is one example that fails (in SemidefiniteProgramming):
```
-- -*- M2-comint -*- hash: -236116570
Solvers configured: CSDP
Default solver: CSDP

i1 : P = sdp(matrix{{1,0},{0,2}}, matrix{{0,1},{1,0}}, matrix{{-1}})

o1 = SDP{A => 1 : (| 0 1 |)}
                   | 1 0 |
         b => | -1 |
         C => | 1 0 |
              | 0 2 |

o1 : SDP

i2 : (X,y,Z,stat) = optimize P;
stdio:2:16:(3): error: Command could not be executed.
```

When I run it in M2 at the command line, it runs fine.

mikestillman commented:
@mahrud , Surprisingly (given the problems with building examples) `ctest -R check-SumsOfSquares -V` (and similarly with SemidefiniteProgramming) come back with no errors.

mikestillman commented:
I wonder what is different in running tests vs the examples, and why it seems to be able to find the executables in the tests, as well as directly executed in M2 by hand, but not in the examples...?

mahrud commented on M2/Macaulay2/tests/engine/raw-ring2.m2:
```diff
@@ -6,7 +6,7 @@
 -----------------------------
 needs "raw-util.m2"
 mo = rawMonomialOrdering { GRevLex => {1,2,3,4} }
-M = rawMonoid(mo, (a,b,c,d)/toString, degring 2, (0,1, 0,1, 1,0, 1,0))
```
But the last change to rawMonoid (in interface.dd) was in d6b1b11bae from 2010! So I kinda think the tests have been ignored for some time.

DanGrayson replied:
Could be!  Maybe those tests became obsolete after the unit tests were implemented.

mahrud replied:
The unit tests aren't as extensive, and don't test the same thing; e.g. the one you highlighted is testing something in interface.dd, but nothing in unit tests does that.

DanGrayson commented on M2/Macaulay2/m2/startup.m2.in:
```diff
@@ -398,6 +398,8 @@ usage := arg -> (
      << "    --stop             exit on error" << newline
      << "    --texmacs          TeXmacs session mode" << newline
      << "    --version          print version number and exit" << newline
+     << "    --webapp           WebApp session mode" << newline
```
It's probably better to merge "master" into your branch rather than "development", because "master" has been tested overnight.

mahrud replied:
But you set the target of this branch as development, which caused a merge conflict. How should we resolve this?

DanGrayson replied:
Well, hmm, one could first try merging with "master" -- that might resolve the conflict.  If not and the conflict is provoked by a recent pull request from someone else, wait a day and see if it finds its way into "master" or gets reverted, due to failing the overnight test.

mahrud commented:
@mikestillman Oh you said failed examples, sorry I was confused. Hmm, I still don't get an error on ubuntu 18.04 running in docker. Your log says CSDP is configured, so I don't think it's not finding it. Can you set Verbosity=1 and run that code again?

mikestillman commented:
Using the cmake branch, building with configure, on ubuntu 18.04, I get the following error during build:

```
CXX interp-tmp.cc
: making startup.c
CC startup.c
g++-8 -std=gnu++14 -g3 -O2  -std=gnu++11  -pipe -Wall -Wno-shadow -Wcast-qual -Wno-sign-conversion -Wno-sign-compare -Wno-parentheses -Wno-sign-compare  -fno-stack-protector -W\
uninitialized -I/usr/include -D__FFLASFFPACK_HAVE_CBLAS -g -O2 -fdebug-prefix-map=/build/fflas-ffpack-Fu5RpH/fflas-ffpack-2.2.2=. -fstack-protector-strong -Wformat -Werror=form\
at-security -fabi-version=6 -std=gnu++11 -fopenmp -Wno-error=pragmas -Wreturn-type -Wunused-function -Wfatal-errors -Wno-unknown-pragmas -Wno-mismatched-tags -I. -I../../../../\
../../Macaulay2/d -I../../../../../../Macaulay2/d/../e -I../../../../../../Macaulay2/d/../system -I../../../../../../Macaulay2/d/../../include -I../../../../../../Macaulay2/d/.\
./c -I../../../../../../Macaulay2/d/../regex -I/home/mike/src/M2-cmake/M2/BUILD/mike/builds.tmp/opt64-gcc8/../../../../include -I/home/mike/src/M2-cmake/M2/BUILD/mike/builds.tm\
p/opt64-gcc8/include -I/home/mike/src/M2-cmake/M2/BUILD/mike/builds.tmp/opt64-gcc8/usr-host/include -I/home/mike/src/M2-cmake/M2/BUILD/mike/builds.tmp/opt64-gcc8/usr-host/inclu\
de/cdd  -I/home/mike/src/M2-cmake/M2/BUILD/mike/builds.tmp/opt64-gcc8/../../../../submodules/memtailor/src  -I/home/mike/src/M2-cmake/M2/BUILD/mike/builds.tmp/opt64-gcc8/../../\
../../submodules/mathic/src  -I/home/mike/src/M2-cmake/M2/BUILD/mike/builds.tmp/opt64-gcc8/../../../../submodules/mathicgb/src -isystem /usr/include/libxml2 -isystem /usr/inclu\
de/libxml2      -DNDEBUG  -I/usr/include/eigen3 -DNDEBUG -Wno-unknown-pragmas  -c -o main.o ../../../../../../Macaulay2/d/main.cpp
../../../../../../Macaulay2/d/main.cpp: In function ‘void* profFunc(ArgCell*)’:
../../../../../../Macaulay2/d/main.cpp:142:24: error: ‘chrono_literals’ is not a namespace-name
   using namespace std::chrono_literals;
                        ^~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
<builtin>: recipe for target 'main.o' failed
make[2]: *** [main.o] Error 1
make[2]: Leaving directory '/home/mike/src/M2-cmake/M2/BUILD/mike/builds.tmp/opt64-gcc8/Macaulay2/d'
Makefile:14: recipe for target 'all-in-d' failed
make[1]: *** [all-in-d] Error 2
make[1]: Leaving directory '/home/mike/src/M2-cmake/M2/BUILD/mike/builds.tmp/opt64-gcc8/Macaulay2'
GNUmakefile:210: recipe for target 'all-in-Macaulay2' failed
make: *** [all-in-Macaulay2] Error 2
make: Leaving directory '/home/mike/src/M2-cmake/M2/BUILD/mike/builds.tmp/opt64-gcc8'
```

mikestillman commented:
> @mikestillman Oh you said failed examples, sorry I was confused. Hmm, I still don't get an error on ubuntu 18.04 running in docker. Your log says CSDP is configured, so I don't think it's not finding it. Can you set Verbosity=1 and run that code again?

@mahrud Is this a cmake, ninja or M2 flag?

mahrud commented:
@mikestillman There are two `-std=gnu++11` in the compile command. That's what's messing it up. Did you build the libraries fresh? Are you using the system fflas_ffpack? That one and givaro are usually the culprit, since the configure build adds their cflags `-std=gnu++11` among them, to our cflags. However, if you build fflas_ffpack with configure, they add `-std=gnu++14`, because I changed that in `configure.ac`

mahrud commented:
@mikestillman ah, sorry, Verbosity is an export flag from SemidefiniteProgramming. Not sure where you would add this, maybe `-e Verbosity=1` to the command line for installing the package?

mikestillman commented:
I added it into one of the failing examples.  When I do that, all examples in that package seem to work.

How do I re run the build of examples for a package?

> On May 30, 2020, at 7:15 PM, Mahrud Sayrafi <notifications@github.com> wrote:
> 
> 
> @mikestillman <https://github.com/mikestillman> ah, sorry, Verbosity is an export flag from SemidefiniteProgramming. Not sure where you would add this, maybe -e Verbosity=1 to the command line for installing the package?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Macaulay2/M2/pull/1190#issuecomment-636396879>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAY6R74WKO54QPCWLQCPC3RUGHRHANCNFSM4NIEYJFQ>.
>

mahrud commented:
Try `ninja uninstall-SemidefiniteProgramming` then `ninja install-SemidefiniteProgramming`.

You can also do:
```
RerunExamples=true cmake .
RerunExamples=true ninja install-Semidefinite
```
Setting it before ninja is only because sometimes ninja reruns cmake, and you don't want that to unset the parameter. Or alternatively in one line:
```
RerunExamples=true cmake --build . --target install-Semidefinite
```

mikestillman commented:
> Try `ninja uninstall-SemidefiniteProgramming` then `ninja install-SemidefiniteProgramming`.
> 
> You can also do:
> 
> ```
> RerunExamples=true cmake .
> RerunExamples=true ninja install-Semidefinite
> ```
> 
> Setting it before ninja is only because sometimes ninja reruns cmake, and you don't want that to unset the parameter. Or alternatively in one line:
> 
> ```
> RerunExamples=true cmake --build . --target install-Semidefinite
> ```

Thanks @mahrud.  In the meantime, I think I did a pull from your github.  Not sure if this has anything to do with the change of behavior I am now seeing!  (Possibly it does).

When I rerun examples (now from a re-checked out SemidefiniteProgramming and SumsOfSquares), those examples all pass fine.

I am rerunning `ninja install-packages` (after setting RemakeAllDocumentation=true).  We will see if that is OK too.

mikestillman commented:
> @mikestillman There are two `-std=gnu++11` in the compile command. That's what's messing it up. Did you build the libraries fresh? Are you using the system fflas_ffpack? That one and givaro are usually the culprit, since the configure build adds their cflags `-std=gnu++11` among them, to our cflags. However, if you build fflas_ffpack with configure, they add `-std=gnu++14`, because I changed that in `configure.ac`

@mahrud: what versions are we using for ffpack and givaro in the cmake build?  Are we building both as submodules? How about in the configure build?

mahrud commented:
@mikestillman fflas is the same version, and both install 2.4.3 from a submodule, I believe. CMake uses Givaro 4.1.1 from a submodule, autotools downloads a tarfile of the same version. They both apply the same patch.

For the isUnit/isunit error, I tried to make it work with either newer or older versions of Givaro with this:
https://github.com/mahrud/M2/blob/feature/cmake/M2/Macaulay2/e/aring-gf-givaro.cpp#L410-L417
But if we can require 4.0.3 or above in configure.ac, I can remove that and remove HAVE_GIVARO_isunit from CMake and configure.ac. I think in your case the error is caused because configure.ac sees the system version of givaro which is older and sets isunit, but you're building and linking the newer givaro which has isUnit, so you run into an error. So this bit needs to change:
https://github.com/mahrud/M2/blob/feature/cmake/M2/configure.ac#L869-L888

@DanGrayson could you pull the fflas_ffpack master in [our fork](https://github.com/Macaulay2/fflas-ffpack)? I think it's missing some tags.

DanGrayson commented on M2/Macaulay2/tests/threads/schur-2.m2:
```diff
@@ -0,0 +1,5 @@
+needsPackage "SchurRings"
```
This new test in this new directory should also be made to run from `tests/Makefile.in`.

mahrud replied:
Done.

DanGrayson commented:
> @DanGrayson could you pull the fflas_ffpack master in [our fork](https://github.com/Macaulay2/fflas-ffpack)? I think it's missing some tags.


```
einsteinium$ git push --tags
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To github.com:Macaulay2/fflas-ffpack
 * [new tag]           2.4.3 -> 2.4.3
 * [new tag]           v2.3.0 -> v2.3.0
 * [new tag]           v2.3.1 -> v2.3.1
 * [new tag]           v2.3.2 -> v2.3.2
 * [new tag]           v2.4.0 -> v2.4.0
 * [new tag]           v2.4.1 -> v2.4.1
 * [new tag]           v2.4.2 -> v2.4.2
```

mahrud commented:
@DanGrayson Thanks for pulling fflas_ffpack. Could you pull Singular and givaro as well? Then all upstreams would be Macualay2 forks. Although, I kinda like the "live on the HEAD" philosophy, but applying patches to a fork is easier than keeping patch files around.

mahrud commented:
@DanGrayson continuing the UserMode conversation here, since it's getting hard to find it:
I made a change based on your suggestion to make it more readable. Let me know what you think.

To be clear, I see this as a temporary patch. In the long term I think we should work to avoid running an external M2 process (except in rare situations like the interrupt tests you wrote) and instead use threads available to us, especially for examples. Running tests and examples should take a fraction of what it currently does.

DanGrayson commented on M2/cmake/build-libraries.cmake:
```diff
@@ -0,0 +1,1172 @@
+set(factory_FLINT_HOME_PATH "${M2_HOST_PREFIX} ${FLINT_INCLUDE_DIR}/..")
+ExternalProject_Add(build-factory
+  GIT_REPOSITORY    ${CMAKE_SOURCE_DIR}/submodules/Singular/.git
+  GIT_TAG           1e4e3f39
```
What's this git commit number doing here?  Does it override the commit number recorded by git for the submodule?  If so, that seems to be a recipe for causing a discrepancy between the cmake build and the autoconf build, and probably could be fixed by simply deleting the for git_tag lines in this file.  (Mike noticed this.  @mikestillman )

mahrud replied:
@DanGrayson Currently, Factory, Givaro, and fflas-ffpack dont support out of tree builds, so the solution here has been to use the submodule *as a remote* and make a new, clean clone in the build directory, then build directly there. I'll replace the commit number with HEAD, which I believe would then use the checked out commit from from the submodule.

DanGrayson replied:
Frobby, too, right?

mahrud replied:
Yes, I forgot that. I replaced all of them with HEAD. Also, I removed the Singular submodule, since the tarfile for 4.1.3 apparently doesn't include the extra `factory-4.1.3` directory, so it is definitely simpler. I also added my patch for it, so you can use that with autotools as well.

DanGrayson commented:
> @DanGrayson continuing the UserMode conversation here, since it's getting hard to find it:
> I made a change based on your suggestion to make it more readable. Let me know what you think.

I think it's worse: there is no reason to introduce the synonym ArgNoReadline for "--no-readline", because "--no-readline" is already perfectly readable.

> 
> To be clear, I see this as a temporary patch. In the long term I think we should work to avoid running an external M2 process (except in rare situations like the interrupt tests you wrote) and instead use threads available to us, especially for examples. Running tests and examples should take a fraction of what it currently does.

I've never seen a speed-up using threads of more than 1.4.  We could run them in separate processes and get n-fold speed-up immediately.

DanGrayson commented:
> > To be clear, I see this as a temporary patch. In the long term I think we should work to avoid running an external M2 process (except in rare situations like the interrupt tests you wrote) and instead use threads available to us, especially for examples. Running tests and examples should take a fraction of what it currently does.
> 
> I've never seen a speed-up using threads of more than 1.4. We could run them in separate processes and get n-fold speed-up immediately.

Also, if you run them in threads, you can't impose memory limits or time limits, so that's no good.

mikestillman commented:
> I've never seen a speed-up using threads of more than 1.4. We could run them in separate processes and get n-fold speed-up immediately.

This part at least is definitely something we should try to fix (but that would be another issue).

DanGrayson commented on M2/Macaulay2/m2/html.m2:
```diff
@@ -462,26 +462,80 @@ describeReturnCode = r -> (
+     cmd = cmd | readmode(SetInputFile,   "<" | format inf);
+     cmd = cmd | readmode(SetOutputFile,  ">>" | format toAbsolutePath tmpf);
+     cmd = cmd | readmode(SetCaptureErr,  "2>&1");
+     if debugLevel > 0 then stderr << cmd << endl;
```
Sometimes the user will need the exact command line for debugging purposes, but now it will be hidden, with no documented way, aside from reading the source code, for getting it to be displayed.   If you really want this useful information to be hidden, then it should probably be connected to the Verbose option and documented.

mahrud replied:
set `debugLevel = 1`.

mahrud replied:
As you said:
> Sometimes the user will need the exact command line for **debugging purposes**

So it is clear.

DanGrayson replied:
Well, sure, if the user reads the source code, then they can figure it out.  But what user will be able to do that?  That's a high bar for users.

mahrud replied:
Then that's a problem with the Core that sound be fixed. Increasing `debugLevel` to get debug output is a natural thing to do.

DanGrayson replied:
That idea might not occur to all users, if it's not documented.  Here's an idea: display the command afterwards, if the example or test failed.

mahrud replied:
Done.

mahrud commented:
> I've never seen a speed-up using threads of more than 1.4.

Sounds like the threading system isn't working.

> We could run them in separate processes and get n-fold speed-up immediately.

Have you ever tried this? Multiple things in html.m2 break down if you do that. Not to mention most computers can't handle starting more than a few M2 processes (each of which by default have 16 threads!) working at the same time.

DanGrayson commented:
> > I've never seen a speed-up using threads of more than 1.4.
> 
> Sounds like the threading system isn't working.


Right.  Our theory was that it was due to libgc, and that's why Bill Furnish started his conservative garbage collector project.

> 
> > We could run them in separate processes and get n-fold speed-up immediately.
> 
> Have you ever tried this? Multiple things in html.m2 break down if you do that. Not to mention most computers can't handle starting more than a few M2 processes (each of which by default have 16 threads!) working at the same time.

We can easily arrange to run them all with just one thread.  

What breaks down in html.m2?

mahrud commented:
> The startup time for each M2 instance is pretty dramatic when we are doing it hundreds of times.

2692 tests and ~4000 examples at the moment. Each startup takes about 0.5s for me; that's almost an hour.

DanGrayson commented:
> > The startup time for each M2 instance is pretty dramatic when we are doing it hundreds of times.
> 
> 2692 tests and ~4000 examples at the moment. Each startup takes about 0.5s for me; that's almost an hour.

Here's an idea: start an M2 process and give it 16 tests to run.  After starting up, it forks into 16 processes to run the tests.  That will cut 1 hour to less than 4 minutes immediately.

mahrud commented:
@DanGrayson this thread is getting out of hand and I'm not interested in arguing with you. I left UserMode intact, as you asked.

DanGrayson commented:
This code has passed the overnight test and has been pulled into the branch "master" just now.
